docs: ABCI 2.0 tutorial (#1374)
cometbft/cometbft: docs/tutorials/forum-application/8.app.md

# App.go

*In this section you will find the code for entire `app.go` file.*

```go
package app

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/andynog/abci2-forum-app/model"

	abci "github.com/cometbft/cometbft/abci/types"
	cryptoproto "github.com/cometbft/cometbft/api/cometbft/crypto/v1"
	cryptoencoding "github.com/cometbft/cometbft/crypto/encoding"

	"github.com/cometbft/cometbft/version"
	"github.com/dgraph-io/badger/v3"
)

const ApplicationVersion = 1

type ForumApp struct {
	abci.BaseApplication

> 	valAddrToPubKeyMap map[string]cryptoproto.PublicKey

Reviewer: How do we recover this map is we crash and recover?

Reviewer 2: in `InitChain` there's an `updateValidator` that set's the value in the app state. So if the node crashes it retrieves from the app state in the DB, as long as the app is initialized fine, at least the genesis validator should be there. But we don't have logic to do validator updates in the tutorial. Could be a future enhancement.

	// Verify that we do not have double words and the validator is not trying to cheat us
	for _, word := range curseWords {
		tmpCurseWordMap[word] = struct{}{}
	}
	if len(tmpCurseWordMap) < len(curseWords) {
		// Extension repeats words
		return &abci.VerifyVoteExtensionResponse{Status: abci.VERIFY_VOTE_EXTENSION_STATUS_REJECT}, nil
	}
	return &abci.VerifyVoteExtensionResponse{Status: abci.VERIFY_VOTE_EXTENSION_STATUS_ACCEPT}, nil
}

// State Sync Connection

func (app *ForumApp) ListSnapshots(_ context.Context, _ *abci.ListSnapshotsRequest) (*abci.ListSnapshotsResponse, error) {
	return &abci.ListSnapshotsResponse{}, nil
}

func (app *ForumApp) OfferSnapshot(_ context.Context, _ *abci.OfferSnapshotRequest) (*abci.OfferSnapshotResponse, error) {
	return &abci.OfferSnapshotResponse{}, nil
}

func (app *ForumApp) LoadSnapshotChunk(_ context.Context, _ *abci.LoadSnapshotChunkRequest) (*abci.LoadSnapshotChunkResponse, error) {
	return &abci.LoadSnapshotChunkResponse{}, nil
}

func (app *ForumApp) ApplySnapshotChunk(_ context.Context, _ *abci.ApplySnapshotChunkRequest) (*abci.ApplySnapshotChunkResponse, error) {
	return &abci.ApplySnapshotChunkResponse{}, nil

> }

Reviewer: I think you can remove these, as you embedded `abci.BaseApplication` in the app struct

// getWordsFromVE gets the curse words from the vote extensions

> func (app *ForumApp) getWordsFromVe(voteExtensions []abci.ExtendedVoteInfo) string {

Reviewer: I think this should be explained (as I hasn't in previous sections)

Reviewer 2: This is just a convenience method used in PrepareProposal, I've added more info the the methods comment to make more clear what it does

	curseWordMap := make(map[string]int)
	for _, vote := range voteExtensions {

		// This code gets the curse words and makes sure that we do not add them more than once
		// Thus ensuring each validator only adds one word once
		curseWords := strings.Split(string(vote.GetVoteExtension()), "|")

		for _, word := range curseWords {
			if count, ok := curseWordMap[word]; !ok {
				curseWordMap[word] = 1
			} else {
				curseWordMap[word] = count + 1
			}
		}

	}
	fmt.Println("Processed vote extensions :", curseWordMap)

> 	majority := len(app.valAddrToPubKeyMap) / 3 // We define the majority to be at least 1/3 of the validators;

Reviewer: ```suggestion
	majority := len(app.valAddrToPubKeyMap) / 3 + 1 // We define the majority to be more than 1/3 of the validators;
```
